src/main/java/org/jabref/logic/search/retrieval/LuceneSearcher.java

```diff
@@ -0,0 +1,127 @@
+            entriesMap.put(bibEntry.getId(), bibEntry);
+            for (LinkedFile linkedFile : bibEntry.getFiles()) {
+                linkedFielsMap.computeIfAbsent(linkedFile.getLink(), k -> new ArrayList<>()).add(bibEntry);
+            }
+        }
+
+        Highlighter highlighter = new Highlighter(new SimpleHTMLFormatter("<b>", "</b>"), new QueryScorer(searchQuery.getQuery()));
+        Analyzer analyzer = SearchFieldConstants.ANALYZER;
+
+        long startTime = System.currentTimeMillis();
+        for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+            float luceneScore = scoreDoc.score;
+            String fileLink = getFieldContents(storedFields, scoreDoc, SearchFieldConstants.PATH);
+            if (!fileLink.isEmpty()) {
+                List<BibEntry> entriesWithFile = linkedFielsMap.get(fileLink);
+                if (entriesWithFile.isEmpty()) {
+                    continue;
+                }
+                int pageNumber = Integer.parseInt(getFieldContents(storedFields, scoreDoc, SearchFieldConstants.PAGE_NUMBER));
+                long modified = Long.parseLong(getFieldContents(storedFields, scoreDoc, SearchFieldConstants.MODIFIED));
+                String content = getFieldContents(storedFields, scoreDoc, SearchFieldConstants.CONTENT);
+                String annotations = getFieldContents(storedFields, scoreDoc, SearchFieldConstants.ANNOTATIONS);
+
+                searchResults.addSearchResult(entriesWithFile,
+                        new SearchResult(luceneScore, fileLink, pageNumber, modified,
+                                getHighlighterFragments(highlighter, analyzer, SearchFieldConstants.CONTENT, content),
+                                getHighlighterFragments(highlighter, analyzer, SearchFieldConstants.ANNOTATIONS, annotations)));
+            } else {
+                String entryId = getFieldContents(storedFields, scoreDoc, SearchFieldConstants.BIB_ENTRY_ID);
+                String defaultField = "";
+                searchResults.addSearchResult(entriesMap.get(entryId),
+                        new SearchResult(luceneScore, getHighlighterFragments(highlighter, analyzer, SearchFieldConstants.DEFAULT_FIELD, defaultField)));
+            }
+        }
+        LOGGER.debug("Mapping search results took {} ms", System.currentTimeMillis() - startTime);
+        return searchResults;
+    }
+
+    private static List<String> getHighlighterFragments(Highlighter highlighter, Analyzer analyzer, String field, String content) {
+        try (TokenStream contentStream = analyzer.tokenStream(field, content)) {
+            TextFragment[] frags = highlighter.getBestTextFragments(contentStream, content, true, 10);
+            return Arrays.stream(frags).map(TextFragment::toString).toList();
+        } catch (IOException | InvalidTokenOffsetsException e) {
+            return List.of();
+        }
+    }
+
+    private static String getFieldContents(StoredFields storedFields, ScoreDoc scoreDoc, String field) {
```

Comment:
Please try to avoid `String`. You can use `enum` somehow. Maybe, also `.toString()` if you interact with lucene. But before, one should use enum for type safety.